---

_You are currently looking at **version 1.0** of this notebook. To download notebooks and datafiles, as well as get help on Jupyter notebooks in the Coursera platform, visit the [Jupyter Notebook FAQ](https://www.coursera.org/learn/python-data-analysis/resources/0dhYG) course resource._

---

# The Series Data Structure

**The series is one of the core data structures in pandas. You think of it a cross between a list and a dictionary. The items are all stored in an order and there's labels with which you can retrieve them. An easy way to visualize this is two columns of data. The first is the special index, a lot like the dictionary keys. While the second is your actual data. It's important to note that the data column has a label of its own and can be retrieved using the .name attribute. This is different than with dictionaries and is useful when it comes to merging multiple columns of data. And we'll talk about that later on in the course.**

In [100]:
import pandas as pd
# pd.Series?

**Let's see an example of this. First, I'll start off by importing the pandas library as pd, then let's take a look at the series object. Here you could see the documentation indicates that you can pass in some data, an index and a name. The data can be anything, that's array-like, like a list. So let's give that a try. We'll just make a list of the three of my favorite animals, a tiger, a bear and a moose. We'll see here that the pandas has automatically identified the type of the data being held in the list, in this case we passed in a list of strings and panda set the type to object.** 

**As you might expect, you can create a series by passing in a list of values. When you do this, Pandas automatically assigns an index starting with zero and sets the name of the series to None. We don't have to use strings.**

In [203]:
animals = ['Tiger', 'Bear','Moose']
anim = pd.Series(animals)
anim

0    Tiger
1     Bear
2    Moose
dtype: object

**If we passed in a list of whole numbers, for instance, we could see that panda sets the type to n 64. Underneath panda stores series values in a typed array using the Numpy library. This offers significant speed-up when processing data versus traditional python lists.**

In [102]:
numbers = [1, 2, 3]
pd.Series(numbers)

0    1
1    2
2    3
dtype: int64

**There's some other typing details that exist for performance that are important to know. The most important is how Numpy and thus pandas handle missing data. In Python, we have the none type to indicate a lack of data. But what do we do if we want to have a typed list like we do in the series object? 
Underneath, pandas does some type conversion. If we create a list of strings and we have one element, a None type, pandas inserts it as a None and uses the type object for the underlying array.**

In [103]:
animals = ['Tiger', 'Bear', None]
pd.Series(animals)

0    Tiger
1     Bear
2     None
dtype: object

**If we create a list of numbers, integers or floats, and put in the None type, pandas automatically converts this to a special floating point value designated as NAN, which stands for not a number. 
For those who might not have done scientific computing in Python before, this is a pretty important point.**

In [104]:
numbers = [1, 2, None]
pd.Series(numbers)

0    1.0
1    2.0
2    NaN
dtype: float64

**NAN is not none and when we try the equality test, it's false.**

In [105]:
import numpy as np
np.nan == None

False

** It turns out that you actually can't do an equality test of NAN to itself. When you do, the answer is always false. You need to use special functions to test for the presence of not a number, such as the Numpy library is NAN.**

**We'll come back to this later in the assignment, but keep in mind when you see NAN, it's meaning is similar to none, but it's a numeric value and it's treated differently for efficiency reasons. **

In [106]:
np.nan == np.nan

False

In [107]:
np.isnan(np.nan)

True

** Let's talk more about how pandas' series can be created. While my list of animals might be a common way to create some play data, often you have label data that you want to manipulate. A series can be created from dictionary data. If you do this, the index is automatically assigned to the keys of the dictionary that you provided and not just incrementing integers. Here's an example using some data from Wikipedia on official national sports. When we create the series, we see that, since it was string data, panda set the data type of the series to object. We set the list of the countries as the value of the series and that the index values can be set to the keys from our dictionary.**

In [108]:
sports = {'Archery': 'Bhutan',
          'Golf': 'Scotland',
          'Sumo': 'Japan',
          'Taekwondo': 'South Korea'}
s = pd.Series(sports)
s

Archery           Bhutan
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
dtype: object

**Once the series has been created, we can get the index object using the index attribute. You could also separate your index creation from the data by passing in the index as a list explicitly to the series.**

In [109]:
s.index

Index(['Archery', 'Golf', 'Sumo', 'Taekwondo'], dtype='object')

In [110]:
s = pd.Series(['Tiger', 'Bear', 'Moose'], index=['India', 'America', 'Canada'])
s

India      Tiger
America     Bear
Canada     Moose
dtype: object

** So what happens if your list of values in the index object are not aligned with the keys in your dictionary for creating the series? Well, pandas overrides the automatic creation to favor only and all of the indices values that you provided. So it will ignore it from your dictionary, all keys, which are not in your index, and pandas will add non type or NAN values for any index value you provide, which is not in your dictionary key list. In this example, we pass in a dictionary of four items but only two are preserved in the series object because of the index list. We see that hockey has been added but since it's also in the index list, it has no value associated with it. **

**So that's how we create a series. In the next class, we'll dig deeper and look at how we get data out of the series. **

In [111]:
sports = {'Archery': 'Bhutan',
          'Golf': 'Scotland',
          'Sumo': 'Japan',
          'Taekwondo': 'South Korea'}
s = pd.Series(sports, index=['Golf', 'Sumo', 'Hockey'])
s

Golf      Scotland
Sumo         Japan
Hockey         NaN
dtype: object

# Querying a Series

**A panda.Series can be queried, either by the index position or the index label. As we saw, if you don't give an index to the series, the position and the label are effectively the same values. **
**To query by numeric location, starting at zero, use the iloc attribute. To query by the index label, you can use the loc attribute.** 

** Here's an example using national sporting event data from Wikipedia. Let's say we want to have a list of all of the sports in our index and a list of countries as values. You might keep these in a dictionary and create a series as we discussed.**

In [112]:
sports = {'Archery': 'Bhutan',
          'Golf': 'Scotland',
          'Sumo': 'Japan',
          'Taekwondo': 'South Korea'}
s = pd.Series(sports)
s

Archery           Bhutan
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
dtype: object

** If you wanted to see the fourth country on this, we would use the iloc attribute with the parameter 3**

In [113]:
s.iloc[3]

'South Korea'

**If you wanted to see which country has golf as its national sport, we would use the loc attribute with parameter golf. Keep in mind that iloc and loc are not methods, they are attributes. So you don't use parentheses to query them, but square brackets instead, which we'll call the indexing operator**

** Though in Python, this calls get and set an item methods depending on the context of its use. **

In [114]:
s.loc['Golf']

'Scotland'

** This might seem a bit confusing if you're used to languages where encapsulation of attributes, variables, and properties is common, such as in Java. Pandas tries to make our code a bit more readable and provides a sort of smart syntax using the indexing operator directly on the series itself. For instance, if you pass in an integer parameter, the operator will behave as if you want it to query via the iloc attribute.**  

In [115]:
s[3]

'South Korea'

** If you pass in an object, it will query as if you wanted to use the label based loc attribute.**

In [116]:
s['Golf']

'Scotland'

**So what happens if your index is a list of integers? This is a bit complicated, and Pandas can't determine automatically whether you're intending to query by index position or index label. So you need to be careful when using the indexing operator on the series itself. And the safer option is to be more explicit and use the iloc or loc attributes directly.**

**Here's an example using some new sports data, where countries are indexed by integer. **

In [117]:
sports = {99: 'Bhutan',
          100: 'Scotland',
          101: 'Japan',
          102: 'South Korea'}
s = pd.Series(sports)

**If we try and call s sub zero, we get a key error, because there's no item in the sports list with an index of zero. Instead we have to call iloc explicitly if we want the first item.**

In [118]:
s[0] #This won't call s.iloc[0] as one might expect, it generates an error instead

KeyError: 0

**Okay, so now we know how to get data out of the series. Let's talk about working with the data. A common task is to want to consider all of the values inside of a series and want to do some sort of operation. This could be trying to find a certain number, summarizing data or transforming the data in some way.**

**A typical programmatic approach to this would be to iterate over all the items in the series, and invoke the operation one is interested in. For instance, we could create a data frame of floating point values. Let's think of these as prices for different products.**

In [119]:
s = pd.Series([100.00, 120.00, 101.00, 3.00])
s

0    100.0
1    120.0
2    101.0
3      3.0
dtype: float64

**We could write a little routine which iterates over all of the items in the series and adds them together to get a total. 
This works, but it's slow. Modern computers can do many tasks simultaneously, especially, but not only, tasks involving mathematics.**

In [120]:
total = 0
for item in s:
    total+=item
print(total)

324.0


**Pandas and the underlying NumPy libraries support a method of computation called vectorization.** 

**Vectorization works with most of the functions in the NumPy library, including the sum function.**

**Here's how we would really write the code using the NumPy sum method. First we need to import the numpy module, and then we just call np.sum and pass in an iterable item. In this case, our panda series.**

In [121]:
import numpy as np

total = np.sum(s)
print(total)

324.0


**Now both of these methods create the same value, but is one actually faster? The Jupyter Notebook has a magic function which can help. First, let's create a big series of random numbers. You'll see this used a lot when demonstrating techniques with Pandas.**
**Note that I've just used the head method, which reduces the amount of data printed out by the series to the first five elements.**

In [122]:
#this creates a big series of random numbers
s = pd.Series(np.random.randint(0,1000,10000))
s.head()

0    851
1    440
2     79
3     61
4    813
dtype: int64

**We can actually verify that length of the series is correct using the len function. **

In [124]:
len(s)

10000

**Magic functions begin with a percentage sign. If we type this sign and then hit the Tab key, we can see a list of the available magic functions. You could write your own magic functions too, but that's a little bit outside of the scope of this course. We're actually going to use what's called a cellular magic function. These start with two percentage signs and modify a raptor code in the current Jupyter cell. The function we're going to use is called timeit. And as you may have guessed from the name, this function will run our code a few times to determine, on average, how long it takes. **

**Let's run timeit with our original iterative code. You can give timeit the number of loops that you would like to run. By default, we'll use 1,000 loops. I'll ask timeit here to use 100 runs because we're recording this. 
Not bad. Timeit ran this code and it doesn't seem like it takes very long at all.**

In [130]:
%%timeit -n 100
summary = 0
for item in s:
    summary+=item

709 µs ± 83.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


** Now let's try with vectorization. 
Wow! This is a pretty shocking difference in the speed and demonstrates why data scientists need to be aware of parallel computing features and start thinking in functional programming terms.**

**Related feature in Pandas and NumPy is called broadcasting. With broadcasting, you can apply an operation to every value in the series, changing the series.**

In [131]:
%%timeit -n 100
summary = np.sum(s)

164 µs ± 48.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


**For instance, if we wanted to increase every random variable by 2, we could do so quickly using the += operator directly on the series object. Here I'll just use the head operator to just print out the top five rows in the series.**

In [132]:
s+=2 #adds two to each item in s using broadcasting
s.head()

0    853
1    442
2     81
3     63
4    815
dtype: int64

**The procedural way of doing this would be to iterate through all of the items in the series and increase the values directly. A quick aside here. Pandas does support iterating through a series much like a dictionary, allowing you to unpack values easily. But if you find yourself iterating through a series, you should question whether you're doing things in the best possible way.**

In [133]:
for label, value in s.iteritems():
    s.set_value(label, value+2)
s.head()

/Users/GeYu/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  


0    855
1    444
2     83
3     65
4    817
dtype: int64

In [134]:
s.iteritems?

Signature: s.iteritems()
Docstring: Lazily iterate over (index, value) tuples
File:      ~/anaconda3/lib/python3.6/site-packages/pandas/core/series.py
Type:      method


**Here's how we would do this using the series set value method. 
Let's try and time the two approaches.**

In [140]:
# %%timeit -n 10
# s = pd.Series(np.random.randint(0,1000,10000))
# for label, value in s.iteritems():
#     s.loc[label]= value+2

**Amazing. Not only is it significantly faster, but it's more concise and maybe even easier to read too. The typical mathematical operations you would expect are vectorized, and the NumPy documentation outlines what it takes to create vectorized functions of your own. **

In [138]:
%%timeit -n 10
s = pd.Series(np.random.randint(0,1000,10000))
s+=2

381 µs ± 127 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


**One last note on using the indexing operators to access series data. The .loc attribute lets you not only modify data in place, but also add new data as well. If the value you pass in as the index doesn't exist, then a new entry is added. And keep in mind, indices can have mixed types. While it's important to be aware of the typing going on underneath, Pandas will automatically change the underlying NumPy types as appropriate. **

**Here's an example using a series of a few numbers. We could add some new value, maybe an animal, as you know, I like bears. Just by calling the .loc indexing operator. We see that mixed types for data values or index labels are no problem for Pandas. **

In [141]:
s = pd.Series([1, 2, 3])
s.loc['Animal'] = 'Bears'
s

0             1
1             2
2             3
Animal    Bears
dtype: object

**Up until now I've shown only examples of a series where the index values were unique. I want to end this lecture by showing an example where index values are not unique, and this makes data frames different, conceptually, that a relational database might be. **

** Revisiting the issue of countries and their national sports, it turns out that many countries seem to like this game cricket. We go back to our original series on sports. It's possible to create a new series object with multiple entries for cricket, and then use append to bring these together. There are a couple of important considerations when using append. First, Pandas is going to take your series and try to infer the best data types to use. In this example, everything is a string, so there's no problems here. **

**Second, the append method doesn't actually change the underlying series. It instead returns a new series which is made up of the two appended together. We can see this by going back and printing the original series of values and seeing that they haven't changed.** 

** This is actually a significant issue for new Pandas users who are used to objects being changed in place. So watch out for it, not just with append but with other Pandas functions as well.** 

In [142]:
original_sports = pd.Series({'Archery': 'Bhutan',
                             'Golf': 'Scotland',
                             'Sumo': 'Japan',
                             'Taekwondo': 'South Korea'})
cricket_loving_countries = pd.Series(['Australia',
                                      'Barbados',
                                      'Pakistan',
                                      'England'], 
                                   index=['Cricket',
                                          'Cricket',
                                          'Cricket',
                                          'Cricket'])
all_countries = original_sports.append(cricket_loving_countries)

In [143]:
original_sports

Archery           Bhutan
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
dtype: object

In [144]:
cricket_loving_countries

Cricket    Australia
Cricket     Barbados
Cricket     Pakistan
Cricket      England
dtype: object

In [145]:
all_countries

Archery           Bhutan
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
Cricket        Australia
Cricket         Barbados
Cricket         Pakistan
Cricket          England
dtype: object

**Finally, we see that when we query the appended series for those who have cricket as their national sport, we don't get a single value, but a series itself. This is actually very common, and if you have a relational database background, this is very similar to every table query resulting in a return set which itself is a table. **

In [146]:
all_countries.loc['Cricket']

Cricket    Australia
Cricket     Barbados
Cricket     Pakistan
Cricket      England
dtype: object

**In this lecture, we focused on one of the primary data types of the Pandas library, the series. There are many more methods associated with this series object that we haven't talked about. But with these basics down, we'll move on to talking about the Panda's two-dimensional data structure, the data frame. The data frame is very similar to the series object, but includes multiple columns of data, and is the structure that you'll spend the majority of your time working with when cleaning and aggregating data. **

** But before we move on to a discussion of the data frame, our tutorial assistant for this course will provide a more in-depth look at how the series object can be used.**

# The DataFrame Data Structure

**The DataFrame data structure is the heart of the Panda's library. It's a primary object that you'll be working with in data analysis and cleaning tasks.**

**The DataFrame is conceptually a two-dimensional series object, where there's an index and multiple columns of content, with each column having a label. In fact, the distinction between a column and a row is really only a conceptual distinction. And you can think of the DataFrame itself as simply a two-axes labeled array.**

**You can create a DataFrame in many different ways, some of which you might expect. For instance, you can use a group of series, where each series represents a row of data. Or you could use a group of dictionaries, where each dictionary represents a row of data.**

**Let's take a look at an example. I'm going to create three purchase order records as series objects for a sort of fictional store. Each series has a name of a customer, the string which describes the item being purchased, and the cost of the items. I like dogs, so I'll purchase a bunch of dog food. Kevin Cullens Thompson, the instructor for the third course in the series, he seems more like a cat man to me, so I'll have him purchase some kitty litter. And I think Vinod, who teaches the fourth course in this series, is more of a bird man, so I'll add some bird seed there.**

**Then we'll feed this into the DataFrame as the first argument and set index values indicating which store where each purchase was done. You'll see that when we print out a DataFrame, the Jupiter notebook's trying to pretty it up a bit, and print it out as a table, which is nice.**

In [147]:
import pandas as pd
purchase_1 = pd.Series({'Name': 'Chris',
                        'Item Purchased': 'Dog Food',
                        'Cost': 22.50})
purchase_2 = pd.Series({'Name': 'Kevyn',
                        'Item Purchased': 'Kitty Litter',
                        'Cost': 2.50})
purchase_3 = pd.Series({'Name': 'Vinod',
                        'Item Purchased': 'Bird Seed',
                        'Cost': 5.00})
df = pd.DataFrame([purchase_1, purchase_2, purchase_3], index=['Store 1', 'Store 1', 'Store 2'])
df.head()

Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5
Store 2  Vinod      Bird Seed   5.0

**Similar to the series, we can extract data using the iLock and Lock attributes. Because the DataFrame is two-dimensional, passing a single value to the lock indexing operator will return series if there's only one row to return. 
In this example, if we wanted to select data associated with Store 2, we would just query the lock attribute with one parameter. 
You'll note that the name of the series is returned as the row index value, while the column name is included in the output as well.**

In [148]:
df.loc['Store 2']

Name                  Vinod
Item Purchased    Bird Seed
Cost                      5
Name: Store 2, dtype: object

**We can check the data type of the return using the python type function. **

In [149]:
type(df.loc['Store 2'])

pandas.core.series.Series

**It's important to remember that the indices and column names along either axes, horizontal or vertical, could be non-unique. For instance, in this example, we see two purchase records for Store 1 as different rows. If we use a single value with the DataFrame lock attribute, multiple rows of the DataFrame will return, not as a new series, but as a new DataFrame. 
For instance, if we query for Store 1 records, we see that Chris and Kevin both shop at the same pets supply store.**

In [57]:
df.loc['Store 1']

Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5

**One of the powers of the Panda's DataFrame is that you can quickly select data based on multiple axes. For instance, if you wanted to just list the costs for Store 1, you would supply two parameters to .log, one being the row index and the other being the column name. If we're only interested in Store 1 costs, we could write this as df.lock('Store 1', 'Cost').**

In [150]:
df.loc['Store 1', 'Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

**What if we just wanted to do column selection and just get a list of all of the costs? Well, there's a couple of options. First, you can get a transpose of the DataFrame, using the capital T attribute, which swaps all of the columns and rows. This essentially turns your column names into indices. **

In [151]:
df.T

Store 1       Store 1    Store 2
Name               Chris         Kevyn      Vinod
Item Purchased  Dog Food  Kitty Litter  Bird Seed
Cost                22.5           2.5          5

**And we can then use the .lock method. This works, but it's pretty ugly.**

In [152]:
df.T.loc['Cost']

Store 1    22.5
Store 1     2.5
Store 2       5
Name: Cost, dtype: object

**Since iloc and loc are used for row selection, the Panda's developers reserved indexing operator directly on the DataFrame for column selection. In a Panda's DataFrame, columns always have a name. So this selection is always label based, not as confusing as it was when using the square bracket operator on the series objects. For those familiar with relational databases, this operator is analogous to column projection. **

In [154]:
df['Cost']

Store 1    22.5
Store 1     2.5
Store 2     5.0
Name: Cost, dtype: float64

**Finally, since the result of using the indexing operators, the DataFrame or series, you can chain operations together. For instance, we could have rewritten the query for all Store 1 costs as df.loc('Store 1', 'Cost'). 
This looks pretty reasonable and gets us the result we wanted.**

**But chaining can come with some costs and is best avoided if you can use another approach. In particular, chaining tends to cause Pandas to return a copy of the DataFrame instead of a view on the DataFrame. 
For selecting a data, this is not a big deal, though it might be slower than necessary. If you are changing data though, this is an important distinction and can be a source of error. **

In [155]:
df.loc['Store 1']['Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

**Here's another method. As we saw, .loc does row selection, and it can take two parameters, the row index and the list of column names. .loc also supports slicing. If we wanted to select all rows, we can use a column to indicate a full slice from beginning to end. And then add the column name as the second parameter as a string. In fact, if we wanted to include multiply columns, we could do so in a list. And Pandas will bring back only the columns we have asked for.**

**Here's an example, where we ask for all of the name and cost values for all stores using the .loc operator.**

**So that's selecting and projecting data from a DataFrame based on row and column labels. The key concepts to remember are that the rows and columns are really just for our benefit. Underneath this is just a two axes labeled array, and transposing the columns is easy.**

**Also, consider the issue of chaining carefully, and try to avoid it, it can cause unpredictable results. Where your intent was to obtain a view of the data, but instead Pandas returns to you a copy. In the Panda's world, friends don't let friends chain calls. So if you see it, point it out, and share a less ambiguous solution.** 

In [156]:
df.loc[:,['Name', 'Cost']]

Name  Cost
Store 1  Chris  22.5
Store 1  Kevyn   2.5
Store 2  Vinod   5.0

**Now, before we leave the discussion of accessing data in DataFrames, let's talk about dropping data.**

**It's easy to delete data in series and DataFrames, and we can use the drop function to do so. This function takes a single parameter, which is the index or row label, to drop. This is another tricky place for new users to pad this. The drop function doesn't change the DataFrame by default. And instead, returns to you a copy of the DataFrame with the given rows removed.**

In [157]:
df.drop('Store 1')

Name Item Purchased  Cost
Store 2  Vinod      Bird Seed   5.0

**We can see that our original DataFrame is still intact.**

In [163]:
df

Name Item Purchased  Cost Location
Store 1  Chris       Dog Food  22.5     None
Store 1  Kevyn   Kitty Litter   2.5     None
Store 2  Vinod      Bird Seed   5.0     None

**Let's make a copy with the copy method and do a drop on it instead. This is a very typical pattern in Pandas, where in place changes to a DataFrame are only done if need be, usually on changes involving indices. So it's important to be aware of.**

**Drop has two interesting optional parameters. The first is called in place, and if it's set to true, the DataFrame will be updated in place, instead of a copy being returned. The second parameter is the axes, which should be dropped. By default, this value is 0, indicating the row axes. But you could change it to 1 if you want to drop a column.**

In [159]:
copy_df = df.copy()
copy_df = copy_df.drop('Store 1')
copy_df

Name Item Purchased  Cost
Store 2  Vinod      Bird Seed   5.0

**There is a second way to drop a column, however. And that's directly through the use of the indexing operator, using the del keyword. 
This way of dropping data, however, takes immediate effect on the DataFrame and does not return a view.**

In [160]:
# copy_df.drop?

In [161]:
del copy_df['Name']
copy_df

Item Purchased  Cost
Store 2      Bird Seed   5.0

**Finally, adding a new column to the DataFrame is as easy as assigning it to some value. For instance, if we wanted to add a new location as a column with default value of none, we could do so by using the assignment operator after the square brackets. This broadcasts the default value to the new column immediately.**

In [162]:
df['Location'] = None
df

Name Item Purchased  Cost Location
Store 1  Chris       Dog Food  22.5     None
Store 1  Kevyn   Kitty Litter   2.5     None
Store 2  Vinod      Bird Seed   5.0     None

# Dataframe Indexing and Loading

**The common work flow is to read your data into a DataFrame then reduce this DataFrame to the particular columns or rows that you're interested in working with. As you've seen, the Panda's toolkit tries to give you views on a DataFrame. This is much faster than copying data and much more memory efficient too.** 

**But it does mean that if you're manipulating the data you have to be aware that any changes to the DataFrame you're working on may have an impact on the base data frame you used originally.**

**Here's an example using our same purchasing DataFrame from earlier. We can create a series based on just the cost category using the square brackets.** 

In [164]:
costs = df['Cost']
costs

Store 1    22.5
Store 1     2.5
Store 2     5.0
Name: Cost, dtype: float64

**Then we can increase the cost in this series using broadcasting. **

In [165]:
costs+=2
costs

Store 1    24.5
Store 1     4.5
Store 2     7.0
Name: Cost, dtype: float64

**Now if we look at our original DataFrame, we see those costs have risen as well. This is an important consideration to watch out for. If you want to explicitly use a copy, then you should consider calling the copy method on the DataFrame for it first. **

In [166]:
df

Name Item Purchased  Cost Location
Store 1  Chris       Dog Food  24.5     None
Store 1  Kevyn   Kitty Litter   4.5     None
Store 2  Vinod      Bird Seed   7.0     None

**In this course, we'll be largely using smaller, moderate-sized datasets. As I mentioned, a common workflow is to read the dataset in, usually from some external file. We saw previously how you can do this using Python, and lists, and dictionaries. You can imagine how you might use those dictionaries to create a Pandas DataFrame.**

**Thankfully, Pandas has built-in support for delimited files such as CSV files as well as a variety of other data formats including relational databases, Excel, and HTML tables.**

**I've saved a CSV file called olympics.csv, which has data from Wikipedia that contains a summary list of the medal various countries have won at the Olympics.**

**We can take a look at this file using the shell command cat. Which we can invoke directly using the exclamation point.**

**What happens here is that when the Jupyter notebook sees a line beginning with an exclamation mark, it sends the rest of the line to the operating system shell for evaluation. So cat works on Linux and Macs, as well as in the Coursera platform, but might not work on Windows. You don't need to worry too much about this. I just wanted to show how a Jupyter notebook can integrate with the operating system to provide you with even more tools to look at your data.**

**We see from the cat output that there seems to be a numeric list of columns followed by a bunch of column identifiers. The column identifiers have some odd looking characters in them. This is the unicode numero sign, which means number of.**

**Then we have rows of data, all columns separated.**


In [167]:
!cat olympics.csv

0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15
,№ Summer,01 !,02 !,03 !,Total,№ Winter,01 !,02 !,03 !,Total,№ Games,01 !,02 !,03 !,Combined total
Afghanistan (AFG),13,0,0,2,2,0,0,0,0,0,13,0,0,2,2
Algeria (ALG),12,5,2,8,15,3,0,0,0,0,15,5,2,8,15
Argentina (ARG),23,18,24,28,70,18,0,0,0,0,41,18,24,28,70
Armenia (ARM),5,1,2,9,12,6,0,0,0,0,11,1,2,9,12
Australasia (ANZ) [ANZ],2,3,4,5,12,0,0,0,0,0,2,3,4,5,12
Australia (AUS) [AUS] [Z],25,139,152,177,468,18,5,3,4,12,43,144,155,181,480
Austria (AUT),26,18,33,35,86,22,59,78,81,218,48,77,111,116,304
Azerbaijan (AZE),5,6,5,15,26,5,0,0,0,0,10,6,5,15,26
Bahamas (BAH),15,5,2,5,12,0,0,0,0,0,15,5,2,5,12
Bahrain (BRN),8,0,0,1,1,0,0,0,0,0,8,0,0,1,1
Barbados (BAR) [BAR],11,0,0,1,1,0,0,0,0,0,11,0,0,1,1
Belarus (BLR),5,12,24,39,75,6,6,4,5,15,11,18,28,44,90
Belgium (BEL),25,37,52,53,142,20,1,1,3,5,45,38,53,56,147
Bermuda (BER),17,0,0,1,1,7,0,0,0,0,24,0,0,1,1
Bohemia (BOH) [BOH] [Z],3,0,1,3,4,0,0,0,0,0,3,0,1,3,4
Botswana (BOT),9,0,1,0,1,0,0,0,0,0,9,0,1,0,1
Brazil (BRA),

**We can read this into a DataFrame by calling the read_csv function of the module. When we look at the DataFrame we see that the first cell has an NaN in it since it's an empty value, and the rows have been automatically indexed for us.**

**It seems pretty clear that the first row of data in the DataFrame is what we really want to see as the column names. It also seems like the first column in the data is the country name, which we would like to make an index.**


In [168]:
df = pd.read_csv('olympics.csv')
df.head()

0         1     2     3     4      5         6     7     8  \
0                NaN  № Summer  01 !  02 !  03 !  Total  № Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  № Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

**Read csv has a number of parameters that we can use to indicate to Pandas how rows and columns should be labeled.** 

**For instance, we can use the index_col to indicate which column should be the index and we can also use the header parameter to indicate which row from the data file should be used as the header. Let's re-import that data and center index value to be 0 which is the first column and let set a column headers to be read from the second row of data. We can do this by using the skip rows parameters, to tell Pandas to ignore the first row, which was made up of numeric column names.** 

In [169]:
df = pd.read_csv('olympics.csv', index_col = 0, skiprows=1)
df.head()

№ Summer  01 !  02 !  03 !  Total  № Winter  01 !.1  \
Afghanistan (AFG)              13     0     0     2      2         0       0   
Algeria (ALG)                  12     5     2     8     15         3       0   
Argentina (ARG)                23    18    24    28     70        18       0   
Armenia (ARM)                   5     1     2     9     12         6       0   
Australasia (ANZ) [ANZ]         2     3     4     5     12         0       0   

                         02 !.1  03 !.1  Total.1  № Games  01 !.2  02 !.2  \
Afghanistan (AFG)             0       0        0       13       0       0   
Algeria (ALG)                 0       0        0       15       5       2   
Argentina (ARG)               0       0        0       41      18      24   
Armenia (ARM)                 0       0        0       11       1       2   
Australasia (ANZ) [ANZ]       0       0        0        2       3       4   

                         03 !.2  Combined total  
Afghanistan (AFG)             2               2  
Algeria (ALG)                 8              15  
Argentina (ARG)              28              70  
Armenia (ARM)                 9              12  
Australasia (ANZ) [ANZ]       5              12

**Now this data came from the all time Olympic games medal table on Wikipedia. If we head to the page we could see that instead of running gold, silver and bronze in the pages, these nice little icons with a one, a two, and a three in them In our csv file these were represented with the strings 01 !, 02 !, and so on. We see that the column values are repeated which really isn't good practice. Panda's recognize this in a panda.1 and .2 to make things more unique.**

**But this labeling isn't really as clear as it could be, so we should clean up the data file. We can of course do this just by going and editing the CSV file directly, but we can also set the column names using the Pandas name property.**

**Panda stores a list of all of the columns in the .columns attribute.** 


In [170]:
df.columns

Index(['№ Summer', '01 !', '02 !', '03 !', 'Total', '№ Winter', '01 !.1',
       '02 !.1', '03 !.1', 'Total.1', '№ Games', '01 !.2', '02 !.2', '03 !.2',
       'Combined total'],
      dtype='object')

**We can change the values of the column names by iterating over this list and calling the rename method of the data frame. Now, I'm going to copy and paste this in to speed up things but we can walk through what is going on here. **

In [171]:
for col in df.columns:
    if col[:2]=='01':
        df.rename(columns={col:'Gold' + col[4:]}, inplace=True)
    if col[:2]=='02':
        df.rename(columns={col:'Silver' + col[4:]}, inplace=True)
    if col[:2]=='03':
        df.rename(columns={col:'Bronze' + col[4:]}, inplace=True)
    if col[:1]=='№':
        df.rename(columns={col:'#' + col[1:]}, inplace=True) 

df.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Afghanistan (AFG)              13     0       0       2      2         0   
Algeria (ALG)                  12     5       2       8     15         3   
Argentina (ARG)                23    18      24      28     70        18   
Armenia (ARM)                   5     1       2       9     12         6   
Australasia (ANZ) [ANZ]         2     3       4       5     12         0   

                         Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
Afghanistan (AFG)             0         0         0        0       13       0   
Algeria (ALG)                 0         0         0        0       15       5   
Argentina (ARG)               0         0         0        0       41      18   
Armenia (ARM)                 0         0         0        0       11       1   
Australasia (ANZ) [ANZ]       0         0         0        0        2       3   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)               0         2               2  
Algeria (ALG)                   2         8              15  
Argentina (ARG)                24        28              70  
Armenia (ARM)                   2         9              12  
Australasia (ANZ) [ANZ]         4         5              12

**Here we just iterate through all of the columns looking to see if they start with a 01, 02, 03 or numeric character. If they do, we can call rename and set the column parameters to a dictionary with the keys being the column we want to replace and the value being the new value we want.**

**Here we'll slice some of the old values in two, since we don't want to lose the unique appended values. We'll also set the ever-important in place parameter to true so Pandas knows to update this data frame directly.**

**All right, that's more on the data frame with a focus on the workflow of actually getting data in and into a place where we might want to query it. **

# Querying a DataFrame

Before we talk about how to query data frames, we need to talk about Boolean masking. Boolean masking is the heart of fast and efficient querying in NumPy. It's analogous a bit to masking used in other computational areas. 

A Boolean mask is an array which can be of one dimension like a series, or two dimensions like a data frame, where each of the values in the array are either true or false. This array is essentially overlaid on top of the data structure that we're querying. And any cell aligned with the true value will be admitted into our final result, and any sign aligned with a false value will not. 

Boolean masking is powerful conceptually and is the cornerstone of efficient NumPy and pandas querying. This technique is well used in other areas of computer science, for instance, in graphics. But it doesn't really have an analogue in other traditional relational databases, so I think it's worth pointing out here. Boolean masks are created by applying operators directly to the pandas series or DataFrame objects. For instance, in our Olympics data set, you might be interested in seeing only those countries who have achieved a gold medal at the summer Olympics. To build a Boolean mask for this query, we project the gold column using the indexing operator and apply the greater than operator with a comparison value of zero. This is essentially broadcasting a comparison operator, greater than, with the results being returned as a Boolean series. The resultant series is indexed where the value of each cell is either true or false depending on whether a country has won at least one gold medal, and the index is the country name. 

So this builds us the Boolean mask, which is half the battle. 

In [172]:
df['Gold'] > 0

Afghanistan (AFG)                               False
Algeria (ALG)                                    True
Argentina (ARG)                                  True
Armenia (ARM)                                    True
Australasia (ANZ) [ANZ]                          True
Australia (AUS) [AUS] [Z]                        True
Austria (AUT)                                    True
Azerbaijan (AZE)                                 True
Bahamas (BAH)                                    True
Bahrain (BRN)                                   False
Barbados (BAR) [BAR]                            False
Belarus (BLR)                                    True
Belgium (BEL)                                    True
Bermuda (BER)                                   False
Bohemia (BOH) [BOH] [Z]                         False
Botswana (BOT)                                  False
Brazil (BRA)                                     True
British West Indies (BWI) [BWI]                 False
Bulgaria (BUL) [H]          

**What we want to do next is overlay that mask on the data frame. We can do this using the where function. The where function takes a Boolean mask as a condition, applies it to the data frame or series, and returns a new data frame or series of the same shape. Let's apply this Boolean mask to our Olympics data and create a data frame of only those countries who have won a gold at a summer games.**

**We see that the resulting data frame keeps the original indexed values, and only data from countries that met the condition are retained. All of the countries which did not meet the condition have NaN data instead. This is okay. Most statistical functions built into the data frame object ignore values of NaN.**

In [175]:
only_gold = df.where(df['Gold'] > 0)
only_gold.head(10)

# Summer   Gold  Silver  Bronze  Total  # Winter  \
Afghanistan (AFG)               NaN    NaN     NaN     NaN    NaN       NaN   
Algeria (ALG)                  12.0    5.0     2.0     8.0   15.0       3.0   
Argentina (ARG)                23.0   18.0    24.0    28.0   70.0      18.0   
Armenia (ARM)                   5.0    1.0     2.0     9.0   12.0       6.0   
Australasia (ANZ) [ANZ]         2.0    3.0     4.0     5.0   12.0       0.0   
Australia (AUS) [AUS] [Z]      25.0  139.0   152.0   177.0  468.0      18.0   
Austria (AUT)                  26.0   18.0    33.0    35.0   86.0      22.0   
Azerbaijan (AZE)                5.0    6.0     5.0    15.0   26.0       5.0   
Bahamas (BAH)                  15.0    5.0     2.0     5.0   12.0       0.0   
Bahrain (BRN)                   NaN    NaN     NaN     NaN    NaN       NaN   

                           Gold.1  Silver.1  Bronze.1  Total.1  # Games  \
Afghanistan (AFG)             NaN       NaN       NaN      NaN      NaN   
Algeria (ALG)                 0.0       0.0       0.0      0.0     15.0   
Argentina (ARG)               0.0       0.0       0.0      0.0     41.0   
Armenia (ARM)                 0.0       0.0       0.0      0.0     11.0   
Australasia (ANZ) [ANZ]       0.0       0.0       0.0      0.0      2.0   
Australia (AUS) [AUS] [Z]     5.0       3.0       4.0     12.0     43.0   
Austria (AUT)                59.0      78.0      81.0    218.0     48.0   
Azerbaijan (AZE)              0.0       0.0       0.0      0.0     10.0   
Bahamas (BAH)                 0.0       0.0       0.0      0.0     15.0   
Bahrain (BRN)                 NaN       NaN       NaN      NaN      NaN   

                           Gold.2  Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)             NaN       NaN       NaN             NaN  
Algeria (ALG)                 5.0       2.0       8.0            15.0  
Argentina (ARG)              18.0      24.0      28.0            70.0  
Armenia (ARM)                 1.0       2.0       9.0            12.0  
Australasia (ANZ) [ANZ]       3.0       4.0       5.0            12.0  
Australia (AUS) [AUS] [Z]   144.0     155.0     181.0           480.0  
Austria (AUT)                77.0     111.0     116.0           304.0  
Azerbaijan (AZE)              6.0       5.0      15.0            26.0  
Bahamas (BAH)                 5.0       2.0       5.0            12.0  
Bahrain (BRN)                 NaN       NaN       NaN             NaN

**For instance, if we call the df.count on the only gold data frame, we see that there are 100 countries which have had gold medals awarded at the summer games, while if we call count on the original data frame, we see that there are 147 countries total. **

In [176]:
only_gold['Gold'].count()

100

In [177]:
df['Gold'].count()

147

**Often we want to drop those rows which have no data. To do this, we can use the drop NA function. You can optionally provide drop NA the axes it should be considering. Remember that the axes is just an indicator for the columns or rows and that the default is zero, which means rows. **

In [178]:
only_gold = only_gold.dropna()
only_gold.head()

# Summer   Gold  Silver  Bronze  Total  # Winter  \
Algeria (ALG)                  12.0    5.0     2.0     8.0   15.0       3.0   
Argentina (ARG)                23.0   18.0    24.0    28.0   70.0      18.0   
Armenia (ARM)                   5.0    1.0     2.0     9.0   12.0       6.0   
Australasia (ANZ) [ANZ]         2.0    3.0     4.0     5.0   12.0       0.0   
Australia (AUS) [AUS] [Z]      25.0  139.0   152.0   177.0  468.0      18.0   

                           Gold.1  Silver.1  Bronze.1  Total.1  # Games  \
Algeria (ALG)                 0.0       0.0       0.0      0.0     15.0   
Argentina (ARG)               0.0       0.0       0.0      0.0     41.0   
Armenia (ARM)                 0.0       0.0       0.0      0.0     11.0   
Australasia (ANZ) [ANZ]       0.0       0.0       0.0      0.0      2.0   
Australia (AUS) [AUS] [Z]     5.0       3.0       4.0     12.0     43.0   

                           Gold.2  Silver.2  Bronze.2  Combined total  
Algeria (ALG)                 5.0       2.0       8.0            15.0  
Argentina (ARG)              18.0      24.0      28.0            70.0  
Armenia (ARM)                 1.0       2.0       9.0            12.0  
Australasia (ANZ) [ANZ]       3.0       4.0       5.0            12.0  
Australia (AUS) [AUS] [Z]   144.0     155.0     181.0           480.0

**When you find yourself talking about pandas and saying phrases like, often I want to, it's quite likely the developers have included a shortcut for this common operation. For instance, in this example, we don't actually have to use the where function explicitly. The pandas developers allow the indexing operator to take a Boolean mask as a value instead of just a list of column names. The syntax might look a little messy, especially if you're not used to programming languages with overloaded operators, but the result is that you're able to filter and reduce data frames relatively quickly.**

**Here's a more concise example of how we could query this data frame. You'll notice that there are no NaNs when you query the data frame in this manner. pandas automatically filters out the rows with no values.**

In [179]:
only_gold = df[df['Gold'] > 0]
only_gold.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Algeria (ALG)                    12     5       2       8     15         3   
Argentina (ARG)                  23    18      24      28     70        18   
Armenia (ARM)                     5     1       2       9     12         6   
Australasia (ANZ) [ANZ]           2     3       4       5     12         0   
Australia (AUS) [AUS] [Z]        25   139     152     177    468        18   

                           Gold.1  Silver.1  Bronze.1  Total.1  # Games  \
Algeria (ALG)                   0         0         0        0       15   
Argentina (ARG)                 0         0         0        0       41   
Armenia (ARM)                   0         0         0        0       11   
Australasia (ANZ) [ANZ]         0         0         0        0        2   
Australia (AUS) [AUS] [Z]       5         3         4       12       43   

                           Gold.2  Silver.2  Bronze.2  Combined total  
Algeria (ALG)                   5         2         8              15  
Argentina (ARG)                18        24        28              70  
Armenia (ARM)                   1         2         9              12  
Australasia (ANZ) [ANZ]         3         4         5              12  
Australia (AUS) [AUS] [Z]     144       155       181             480

**One more thing to keep in mind if you're not used to Boolean or bit masking for data reduction. The output of two Boolean masks being compared with logical operators is another Boolean mask. This means that you can chain together a bunch of and/or statements in order to create more complex queries, and the result is a single Boolean mask.**

**For instance, we could create a mask for all of those countries who have received a gold in the summer Olympics and logically order that with all of those countries who have received a gold in the winter Olympics. If we apply this to the data frame and use the length function to see how many rows there are, we see that there are 101 countries which have won a gold metal at some time.**

In [180]:
len(df[(df['Gold'] > 0) | (df['Gold.1'] > 0)])

101

**Another example for fun. Have there been any countries who have only won a gold in the winter Olympics and never in the summer Olympics? Here's one way to answer that.**

**Poor Liechtenstein. Thankfully the Olympics come every four years. I know who I'll be cheering for in 2020 to win their first summer gold.**

**Extremely important, and often an issue for new users, is to remember that each Boolean mask needs to be encased in parenthesis because of the order of operations. This can cause no end of frustration if you're not used to it, so be careful.**

**In this lecture, we took a look at Boolean masking. We didn't have much code to write, but applying masks to data frames is really a core panda's work flow and is worth practicing.**

In [181]:
df[(df['Gold.1'] > 0) & (df['Gold'] == 0)]

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Liechtenstein (LIE)        16     0       0       0      0        18       2   

                     Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  \
Liechtenstein (LIE)         2         5        9       34       2         2   

                     Bronze.2  Combined total  
Liechtenstein (LIE)         5               9

# Indexing Dataframes

As we've seen, both series and DataFrames can have indices applied to them. The index is essentially a row level label, and we know that rows correspond to axis zero. In our Olympics data, we indexed the data frame by the name of the country. Indices can either be inferred, such as when we create a new series without an index, in which case we get numeric values, or they can be set explicitly, like when we use the dictionary object to create the series, or when we loaded data from the CSV file and specified the header. Another option for setting an index is to use the set_index function. This function takes a list of columns and promotes those columns to an index. Set index is a destructive process, it doesn't keep the current index. If you want to keep the current index, you need to manually create a new column and copy into it values from the index attribute. 

In [182]:
df.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Afghanistan (AFG)              13     0       0       2      2         0   
Algeria (ALG)                  12     5       2       8     15         3   
Argentina (ARG)                23    18      24      28     70        18   
Armenia (ARM)                   5     1       2       9     12         6   
Australasia (ANZ) [ANZ]         2     3       4       5     12         0   

                         Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
Afghanistan (AFG)             0         0         0        0       13       0   
Algeria (ALG)                 0         0         0        0       15       5   
Argentina (ARG)               0         0         0        0       41      18   
Armenia (ARM)                 0         0         0        0       11       1   
Australasia (ANZ) [ANZ]       0         0         0        0        2       3   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)               0         2               2  
Algeria (ALG)                   2         8              15  
Argentina (ARG)                24        28              70  
Armenia (ARM)                   2         9              12  
Australasia (ANZ) [ANZ]         4         5              12

**Let's go back to our Olympics DataFrame. Let's say that we don't want to index the DataFrame by countries, but instead want to index by the number of gold medals that were won at summer games. First we need to preserve the country information into a new column. We can do this using the indexing operator or the string that has the column label. Then we can use the set_index to set index of the column to summer gold medal wins.** 

**You'll see that when we create a new index from an existing column it appears that a new first row has been added with empty values. This isn't quite what's happening. And we know this in part because an empty value is actually rendered either as a none or an NaN if the data type of the column is numeric. What's actually happened is that the index has a name. Whatever the column name was in the Jupiter notebook has just provided this in the output. **

In [183]:
df['country'] = df.index
df = df.set_index('Gold')
df.head()

# Summer  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  Bronze.1  \
Gold                                                                          
0           13       0       2      2         0       0         0         0   
5           12       2       8     15         3       0         0         0   
18          23      24      28     70        18       0         0         0   
1            5       2       9     12         6       0         0         0   
3            2       4       5     12         0       0         0         0   

      Total.1  # Games  Gold.2  Silver.2  Bronze.2  Combined total  \
Gold                                                                 
0           0       13       0         0         2               2   
5           0       15       5         2         8              15   
18          0       41      18        24        28              70   
1           0       11       1         2         9              12   
3           0        2       3         4         5              12   

                      country  
Gold                           
0           Afghanistan (AFG)  
5               Algeria (ALG)  
18            Argentina (ARG)  
1               Armenia (ARM)  
3     Australasia (ANZ) [ANZ]

**We can get rid of the index completely by calling the function reset_index. This promotes the index into a column and creates a default numbered index.**

**One nice feature of pandas is that it has the option to do multi-level indexing. This is similar to composite keys in relational database systems. To create a multi-level index, we simply call set index and give it a list of columns that we're interested in promoting to an index. 
Pandas will search through these in order, finding the distinct data and forming composite indices.**

In [184]:
df = df.reset_index()
df.head()

Gold  # Summer  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  \
0     0        13       0       2      2         0       0         0   
1     5        12       2       8     15         3       0         0   
2    18        23      24      28     70        18       0         0   
3     1         5       2       9     12         6       0         0   
4     3         2       4       5     12         0       0         0   

   Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  Combined total  \
0         0        0       13       0         0         2               2   
1         0        0       15       5         2         8              15   
2         0        0       41      18        24        28              70   
3         0        0       11       1         2         9              12   
4         0        0        2       3         4         5              12   

                   country  
0        Afghanistan (AFG)  
1            Algeria (ALG)  
2          Argentina (ARG)  
3            Armenia (ARM)  
4  Australasia (ANZ) [ANZ]

**A good example of this is often found when dealing with geographical data which is sorted by regions or demographics. Let's change data sets and look at some census data for a better example. This data is stored in the file census.csv and comes from the United States Census Bureau. In particular, this is a breakdown of the population level data at the US county level. It's a great example of how different kinds of data sets might be formatted when you're trying to clean them. For instance, in this data set there are two summarized levels, one that contains summary data for the whole country. And one that contains summary data for each state, and one that contains summary data for each county. **

In [185]:
df = pd.read_csv('census.csv')
df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
0        4779736            4780127          4785161     ...        
1          54571              54571            54660     ...        
2         182265             182265           183193     ...        
3          27457              27457            27341     ...        
4          22915              22919            22861     ...        

   RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
0          0.002295         -0.193196          0.381066          0.582002   
1          7.242091         -2.915927         -3.012349          2.265971   
2         14.832960         17.647293         21.845705         19.243287   
3         -4.728132         -2.500690         -7.056824         -3.904217   
4         -5.527043         -5.068871         -6.201001         -0.177537   

   RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
0         -0.467369     1.030015     0.826644     1.383282     1.724718   
1         -2.530799     7.606016    -2.626146    -2.722002     2.592270   
2         17.197872    15.844176    18.559627    22.727626    20.317142   
3        -10.543299    -4.874741    -2.758113    -7.167664    -3.978583   
4          0.177258    -5.088389    -4.363636    -5.403729     0.754533   

   RNETMIG2015  
0     0.712594  
1    -2.187333  
2    18.293499  
3   -10.543299  
4     1.107861  

[5 rows x 100 columns]

**I often find that I want to see a list of all the unique values in a given column. In this DataFrame, we see that the possible values for the sum level are using the unique function on the DataFrame. This is similar to the SQL distinct operator.**

**Here we can run unique on the sum level of our current DataFrame and see that there are only two different values, 40 and 50.**

In [186]:
df['SUMLEV'].unique()

array([40, 50])

**Let's get rid of all of the rows that are summaries at the state level and just keep the county data.**

In [187]:
df=df[df['SUMLEV'] == 50]
df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   
5      50       3         6      1       9  Alabama   Blount County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
1          54571              54571            54660     ...        
2         182265             182265           183193     ...        
3          27457              27457            27341     ...        
4          22915              22919            22861     ...        
5          57322              57322            57373     ...        

   RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
1          7.242091         -2.915927         -3.012349          2.265971   
2         14.832960         17.647293         21.845705         19.243287   
3         -4.728132         -2.500690         -7.056824         -3.904217   
4         -5.527043         -5.068871         -6.201001         -0.177537   
5          1.807375         -1.177622         -1.748766         -2.062535   

   RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
1         -2.530799     7.606016    -2.626146    -2.722002     2.592270   
2         17.197872    15.844176    18.559627    22.727626    20.317142   
3        -10.543299    -4.874741    -2.758113    -7.167664    -3.978583   
4          0.177258    -5.088389    -4.363636    -5.403729     0.754533   
5         -1.369970     1.859511    -0.848580    -1.402476    -1.577232   

   RNETMIG2015  
1    -2.187333  
2    18.293499  
3   -10.543299  
4     1.107861  
5    -0.884411  

[5 rows x 100 columns]

**Also while this data set is interesting for a number of different reasons, let's reduce the data that we're going to look at to just the total population estimates and the total number of births. We can do this by creating a list of column names that we want to keep then project those and assign the resulting DataFrame to our df variable.**

In [188]:
columns_to_keep = ['STNAME',
                   'CTYNAME',
                   'BIRTHS2010',
                   'BIRTHS2011',
                   'BIRTHS2012',
                   'BIRTHS2013',
                   'BIRTHS2014',
                   'BIRTHS2015',
                   'POPESTIMATE2010',
                   'POPESTIMATE2011',
                   'POPESTIMATE2012',
                   'POPESTIMATE2013',
                   'POPESTIMATE2014',
                   'POPESTIMATE2015']
df = df[columns_to_keep]
df.head()

STNAME         CTYNAME  BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
1  Alabama  Autauga County         151         636         615         574   
2  Alabama  Baldwin County         517        2187        2092        2160   
3  Alabama  Barbour County          70         335         300         283   
4  Alabama     Bibb County          44         266         245         259   
5  Alabama   Blount County         183         744         710         646   

   BIRTHS2014  BIRTHS2015  POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  \
1         623         600            54660            55253            55175   
2        2186        2240           183193           186659           190396   
3         260         269            27341            27226            27159   
4         247         253            22861            22733            22642   
5         618         603            57373            57711            57776   

   POPESTIMATE2013  POPESTIMATE2014  POPESTIMATE2015  
1            55038            55290            55347  
2           195126           199713           203709  
3            26973            26815            26489  
4            22512            22549            22583  
5            57734            57658            57673

**The US Census data breaks down estimates of population data by state and county. We can load the data and set the index to be a combination of the state and county values and see how pandas handles it in a DataFrame. We do this by creating a list of the column identifiers we want to have indexed. And then calling set index with this list and assigning the output as appropriate. We see here that we have a dual index, first the state name and then the county name.**

In [189]:
df = df.set_index(['STNAME', 'CTYNAME'])
df.head()

BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
STNAME  CTYNAME                                                          
Alabama Autauga County         151         636         615         574   
        Baldwin County         517        2187        2092        2160   
        Barbour County          70         335         300         283   
        Bibb County             44         266         245         259   
        Blount County          183         744         710         646   

                        BIRTHS2014  BIRTHS2015  POPESTIMATE2010  \
STNAME  CTYNAME                                                   
Alabama Autauga County         623         600            54660   
        Baldwin County        2186        2240           183193   
        Barbour County         260         269            27341   
        Bibb County            247         253            22861   
        Blount County          618         603            57373   

                        POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME  CTYNAME                                                             
Alabama Autauga County            55253            55175            55038   
        Baldwin County           186659           190396           195126   
        Barbour County            27226            27159            26973   
        Bibb County               22733            22642            22512   
        Blount County             57711            57776            57734   

                        POPESTIMATE2014  POPESTIMATE2015  
STNAME  CTYNAME                                           
Alabama Autauga County            55290            55347  
        Baldwin County           199713           203709  
        Barbour County            26815            26489  
        Bibb County               22549            22583  
        Blount County             57658            57673

**An immediate question which comes up is how we can query this DataFrame. For instance, we saw previously that the loc attribute of the DataFrame can take multiple arguments. And it could query both the row and the columns. When you use a MultiIndex, you must provide the arguments in order by the level you wish to query. Inside of the index, each column is called a level and the outermost column is level zero. For instance, if we want to see the population results from Washtenaw County, which is where I live, you'd want to the first argument as the state of Michigan. **

In [190]:
df.loc['Michigan', 'Washtenaw County']

BIRTHS2010            977
BIRTHS2011           3826
BIRTHS2012           3780
BIRTHS2013           3662
BIRTHS2014           3683
BIRTHS2015           3709
POPESTIMATE2010    345563
POPESTIMATE2011    349048
POPESTIMATE2012    351213
POPESTIMATE2013    354289
POPESTIMATE2014    357029
POPESTIMATE2015    358880
Name: (Michigan, Washtenaw County), dtype: int64

**You might be interested in just comparing two counties. For instance, Washtenaw where I live and Wayne County which covers Detroit. To do this, we can pass the loc method, a list of tuples which describe the indices we wish to query. Since we have a MultiIndex of two values, the state and the county, we need to provide two values as each element of our filtering list. **

In [191]:
df.loc[ [('Michigan', 'Washtenaw County'),
         ('Michigan', 'Wayne County')] ]

BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
STNAME   CTYNAME                                                            
Michigan Washtenaw County         977        3826        3780        3662   
         Wayne County            5918       23819       23270       23377   

                           BIRTHS2014  BIRTHS2015  POPESTIMATE2010  \
STNAME   CTYNAME                                                     
Michigan Washtenaw County        3683        3709           345563   
         Wayne County           23607       23586          1815199   

                           POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME   CTYNAME                                                               
Michigan Washtenaw County           349048           351213           354289   
         Wayne County              1801273          1792514          1775713   

                           POPESTIMATE2014  POPESTIMATE2015  
STNAME   CTYNAME                                             
Michigan Washtenaw County           357029           358880  
         Wayne County              1766008          1759335

**Okay so that's how hierarchical indices work in a nutshell. They're a special part of the pandas library which I think can make management and reasoning about data easier. Of course hierarchical labeling isn't just for rows. For example, you can transpose this matrix and now have hierarchical column labels. And projecting a single column which has these labels works exactly the way you would expect it to.**

# Missing values

**We're going to end this week of lecture with a quick discussion of missing values. We've seen a preview of how Pandas handles missing values using the None type and NumPy NaN values. Missing values are pretty common in data cleaning activities. There are couple of caveats and discussion points which we should address.** 

**First, the built in loading from delimited files provides control for missing values in a few ways. The most germane of these, is the na_values list, to indicate other strings which could refer to missing values. Some of my sociologist colleagues for instance, regularly use the value of 99 in binary categories to indicate that there's no value. So this comes in handy. You can also use the na_filter option to turn off white space filtering, if white space is an actual value of interest. But in practice, this is pretty rare. In addition to rules controlling how missing values might be loaded, it's sometimes useful to consider missing values as actually having information. I'll give an example from my own research. I often deal with logs from online learning systems. In particular, I've done a couple of projects looking at video use in lecture capture systems. In these systems it's common for the player for have a heartbeat functionality where playback statistics are sent to the server every so often, maybe every 30 seconds.**

**These heartbeats can get big as they can carry the whole state of the playback system, such as where the video play head is at, where the video size is, which video is being rendered to the screen, how loud the volume is, etc.**

**If we load the data file log.txt, we can see an example of what this might look like. In this data the first column is a timestamp in the Unix epoch format. The next column is the user name followed by a web page they're visiting and the video that they're playing.** 

**Each row of the DataFrame has a playback position. And we can see that as the playback position increases by one, the time stamp increases by about 30 seconds.** 

**Except for user Bob. It turns out that Bob has paused his playback so as time increases the playback position doesn't change. Note too how difficult it is for us to try and derive this knowledge from the data, because it's not sorted by time stamp as one might expect. This is actually not uncommon on systems which have a high degree of parallelism.**

**There are a lot of missing values in the paused and volume columns. It's not efficient to send this information across the network if it hasn't changed. So this particular system just inserts null values into the database if there's no changes.**

In [192]:
df = pd.read_csv('log.csv')
df

time    user          video  playback position paused  volume
0   1469974424  cheryl     intro.html                  5  False    10.0
1   1469974454  cheryl     intro.html                  6    NaN     NaN
2   1469974544  cheryl     intro.html                  9    NaN     NaN
3   1469974574  cheryl     intro.html                 10    NaN     NaN
4   1469977514     bob     intro.html                  1    NaN     NaN
5   1469977544     bob     intro.html                  1    NaN     NaN
6   1469977574     bob     intro.html                  1    NaN     NaN
7   1469977604     bob     intro.html                  1    NaN     NaN
8   1469974604  cheryl     intro.html                 11    NaN     NaN
9   1469974694  cheryl     intro.html                 14    NaN     NaN
10  1469974724  cheryl     intro.html                 15    NaN     NaN
11  1469974454     sue  advanced.html                 24    NaN     NaN
12  1469974524     sue  advanced.html                 25    NaN     NaN
13  1469974424     sue  advanced.html                 23  False    10.0
14  1469974554     sue  advanced.html                 26    NaN     NaN
15  1469974624     sue  advanced.html                 27    NaN     NaN
16  1469974654     sue  advanced.html                 28    NaN     5.0
17  1469974724     sue  advanced.html                 29    NaN     NaN
18  1469974484  cheryl     intro.html                  7    NaN     NaN
19  1469974514  cheryl     intro.html                  8    NaN     NaN
20  1469974754     sue  advanced.html                 30    NaN     NaN
21  1469974824     sue  advanced.html                 31    NaN     NaN
22  1469974854     sue  advanced.html                 32    NaN     NaN
23  1469974924     sue  advanced.html                 33    NaN     NaN
24  1469977424     bob     intro.html                  1   True    10.0
25  1469977454     bob     intro.html                  1    NaN     NaN
26  1469977484     bob     intro.html                  1    NaN     NaN
27  1469977634     bob     intro.html                  1    NaN     NaN
28  1469977664     bob     intro.html                  1    NaN     NaN
29  1469974634  cheryl     intro.html                 12    NaN     NaN
30  1469974664  cheryl     intro.html                 13    NaN     NaN
31  1469977694     bob     intro.html                  1    NaN     NaN
32  1469977724     bob     intro.html                  1    NaN     NaN

**One of the handy functions that Pandas has for working with missing values is the filling function, fillna. This function takes a number or parameters, for instance, you could pass in a single value which is called a scalar value to change all of the missing data to one value. This isn't really applicable in this case, but it's a pretty common use case.**

In [195]:
# df.fillna?

**Next up though is the method parameter. The two common fill values are ffill and bfill. ffill is for forward filling and it updates an na value for a particular cell with the value from the previous row. It's important to note that your data needs to be sorted in order for this to have the effect you might want. Data that comes from traditional database management systems usually has no order guarantee, just like this data. So be careful.** 

**In Pandas we can sort either by index or by values. Here we'll just promote the time stamp to an index then sort on the index. **

**If we look closely at the output though we'll notice that the index isn't really unique. Two users seem to be able to use the system at the same time. Again, a very common case.**

In [194]:
df = df.set_index('time')
df = df.sort_index()
df

user          video  playback position paused  volume
time                                                               
1469974424  cheryl     intro.html                  5  False    10.0
1469974424     sue  advanced.html                 23  False    10.0
1469974454  cheryl     intro.html                  6    NaN     NaN
1469974454     sue  advanced.html                 24    NaN     NaN
1469974484  cheryl     intro.html                  7    NaN     NaN
1469974514  cheryl     intro.html                  8    NaN     NaN
1469974524     sue  advanced.html                 25    NaN     NaN
1469974544  cheryl     intro.html                  9    NaN     NaN
1469974554     sue  advanced.html                 26    NaN     NaN
1469974574  cheryl     intro.html                 10    NaN     NaN
1469974604  cheryl     intro.html                 11    NaN     NaN
1469974624     sue  advanced.html                 27    NaN     NaN
1469974634  cheryl     intro.html                 12    NaN     NaN
1469974654     sue  advanced.html                 28    NaN     5.0
1469974664  cheryl     intro.html                 13    NaN     NaN
1469974694  cheryl     intro.html                 14    NaN     NaN
1469974724  cheryl     intro.html                 15    NaN     NaN
1469974724     sue  advanced.html                 29    NaN     NaN
1469974754     sue  advanced.html                 30    NaN     NaN
1469974824     sue  advanced.html                 31    NaN     NaN
1469974854     sue  advanced.html                 32    NaN     NaN
1469974924     sue  advanced.html                 33    NaN     NaN
1469977424     bob     intro.html                  1   True    10.0
1469977454     bob     intro.html                  1    NaN     NaN
1469977484     bob     intro.html                  1    NaN     NaN
1469977514     bob     intro.html                  1    NaN     NaN
1469977544     bob     intro.html                  1    NaN     NaN
1469977574     bob     intro.html                  1    NaN     NaN
1469977604     bob     intro.html                  1    NaN     NaN
1469977634     bob     intro.html                  1    NaN     NaN
1469977664     bob     intro.html                  1    NaN     NaN
1469977694     bob     intro.html                  1    NaN     NaN
1469977724     bob     intro.html                  1    NaN     NaN

**Let's reset the index, and use some multi-level indexing instead, and promote the user name to a second level of the index to deal with that issue. **

In [196]:
df = df.reset_index()
df = df.set_index(['time', 'user'])
df

video  playback position paused  volume
time       user                                                   
1469974424 cheryl     intro.html                  5  False    10.0
           sue     advanced.html                 23  False    10.0
1469974454 cheryl     intro.html                  6    NaN     NaN
           sue     advanced.html                 24    NaN     NaN
1469974484 cheryl     intro.html                  7    NaN     NaN
1469974514 cheryl     intro.html                  8    NaN     NaN
1469974524 sue     advanced.html                 25    NaN     NaN
1469974544 cheryl     intro.html                  9    NaN     NaN
1469974554 sue     advanced.html                 26    NaN     NaN
1469974574 cheryl     intro.html                 10    NaN     NaN
1469974604 cheryl     intro.html                 11    NaN     NaN
1469974624 sue     advanced.html                 27    NaN     NaN
1469974634 cheryl     intro.html                 12    NaN     NaN
1469974654 sue     advanced.html                 28    NaN     5.0
1469974664 cheryl     intro.html                 13    NaN     NaN
1469974694 cheryl     intro.html                 14    NaN     NaN
1469974724 cheryl     intro.html                 15    NaN     NaN
           sue     advanced.html                 29    NaN     NaN
1469974754 sue     advanced.html                 30    NaN     NaN
1469974824 sue     advanced.html                 31    NaN     NaN
1469974854 sue     advanced.html                 32    NaN     NaN
1469974924 sue     advanced.html                 33    NaN     NaN
1469977424 bob        intro.html                  1   True    10.0
1469977454 bob        intro.html                  1    NaN     NaN
1469977484 bob        intro.html                  1    NaN     NaN
1469977514 bob        intro.html                  1    NaN     NaN
1469977544 bob        intro.html                  1    NaN     NaN
1469977574 bob        intro.html                  1    NaN     NaN
1469977604 bob        intro.html                  1    NaN     NaN
1469977634 bob        intro.html                  1    NaN     NaN
1469977664 bob        intro.html                  1    NaN     NaN
1469977694 bob        intro.html                  1    NaN     NaN
1469977724 bob        intro.html                  1    NaN     NaN

** Now that we have the data indexed and sorted appropriately, we can fill the missing datas using ffill. It's good to remember when dealing with missing values so you can deal with individual columns or sets of columns by projecting them just as we spoke about earlier. So you don't have to fix all missing values in one command.**

In [198]:
df = df.fillna(method='ffill')
df

video  playback position  paused  volume
time       user                                                    
1469974424 cheryl     intro.html                  5   False    10.0
           sue     advanced.html                 23   False    10.0
1469974454 cheryl     intro.html                  6   False    10.0
           sue     advanced.html                 24   False    10.0
1469974484 cheryl     intro.html                  7   False    10.0
1469974514 cheryl     intro.html                  8   False    10.0
1469974524 sue     advanced.html                 25   False    10.0
1469974544 cheryl     intro.html                  9   False    10.0
1469974554 sue     advanced.html                 26   False    10.0
1469974574 cheryl     intro.html                 10   False    10.0
1469974604 cheryl     intro.html                 11   False    10.0
1469974624 sue     advanced.html                 27   False    10.0
1469974634 cheryl     intro.html                 12   False    10.0
1469974654 sue     advanced.html                 28   False     5.0
1469974664 cheryl     intro.html                 13   False     5.0
1469974694 cheryl     intro.html                 14   False     5.0
1469974724 cheryl     intro.html                 15   False     5.0
           sue     advanced.html                 29   False     5.0
1469974754 sue     advanced.html                 30   False     5.0
1469974824 sue     advanced.html                 31   False     5.0
1469974854 sue     advanced.html                 32   False     5.0
1469974924 sue     advanced.html                 33   False     5.0
1469977424 bob        intro.html                  1    True    10.0
1469977454 bob        intro.html                  1    True    10.0
1469977484 bob        intro.html                  1    True    10.0
1469977514 bob        intro.html                  1    True    10.0
1469977544 bob        intro.html                  1    True    10.0
1469977574 bob        intro.html                  1    True    10.0
1469977604 bob        intro.html                  1    True    10.0
1469977634 bob        intro.html                  1    True    10.0
1469977664 bob        intro.html                  1    True    10.0
1469977694 bob        intro.html                  1    True    10.0
1469977724 bob        intro.html                  1    True    10.0

** It's sometimes useful to use forward filling, sometimes backwards filling, and sometimes useful to just use a single number. More recently, the Pandas team introduced a method of filling missing values with a series which is the same length as your DataFrame. This makes it easy to derive values which are missing if you have the underlying to do so. For instance, if you're dealing with receipts and you have a column for final price and a column for discount but are missing information from the original price column, you can fill this automatically using fillna.**

**One last note on missing values. When you use statistical functions on DataFrames, these functions typically ignore missing values. For instance if you try and calculate the mean value of a DataFrame, the underlying NumPy function will ignore missing values. This is usually what you want but you should be aware that values are being excluded. **